### DISCUSSION TITLE
🎉 BETA Release: Item Groups, Separators, Line Breaks, Callout Improvements (v1.10)

### DISCUSSION THREAD

**Discussion by chrisgurney** [chrisgurney/obsidian-note-toolbar#114] [Announcements]
The v1.10 release is getting closer to finalization, and I would love some help with testing.

### Installation

> _**Disclaimer**: In this beta release, the format of Note Toolbar's settings has been updated. There may be bugs, so if you wish to give it a go, make a backup of your Note Toolbar `data.json`._

[Install via BRAT.](https://github.com/chrisgurney/obsidian-note-toolbar?tab=readme-ov-file#installation)

- Optional, but would be super helpful: As part of testing, I would _love_ to know if your settings were migrated successfully. As such it would be great to see a version of your `data.json` before and after installation of the beta.
- If your data was not migrated successfully, you may run into [this issue](https://github.com/chrisgurney/obsidian-note-toolbar/issues/112).

### Toolbars within toolbars

An `Item Group` type allows you to directly embed a toolbar within another toolbar.
  - Example:
    - Say Toolbar 1 has these items: `A` `B` `C`, and Toolbar 2 has these items: `D` `E`
    - Add an Item Group type item to the end of Toolbar 2 that's just a reference to Toolbar 1's items: Now Toolbar 2 has these items: `D` `E` `A` `B` `C`
  - Notes:
    - Styling of the "parent" toolbar is maintained. e.g., Toolbar 2 will continue to be styled and positioned per Toolbar 2's settings.
    - Only one level of toolbar is embedded, in an effort to prevent recursion and maintain performance.
    - The icon, label, and tooltip fields are disabled as they don't do anything for item groups.
    - [Variables](https://github.com/chrisgurney/obsidian-note-toolbar/wiki/Variables) used in toolbar items in the embedded toolbar resolve based on the props/values in the current note.

### Line breaks and separators

Support for adding separators and line breaks to your toolbars.
  - Add using the buttons in the lower left of the item list.
  - Separators can be styled via [Style Settings](https://github.com/chrisgurney/obsidian-note-toolbar/wiki/Style-Settings-plugin-support).
  - Breaks are also displayed as separators within menus.
  - Visibility can be set to desktop/mobile/both.

*Example settings showing an item group, a separator, and a line break:*

<img src="https://github.com/user-attachments/assets/73fe7e3f-b1bb-4328-bb31-171453f3ebd6" alt="Settings list of toolbar items that demonstrates separators and line breaks, as well as an item group." width="700"/>

*What the above configured toolbar looks like:*

<img src="https://github.com/user-attachments/assets/449e9efe-8cd0-406a-bbd5-66ffe850d4aa" alt="Toolbar with an item group, line break, and a separator." width="700"/>

### Note Toolbar Callout Improvements

[Note Toolbar Callouts](https://github.com/chrisgurney/obsidian-note-toolbar/wiki/Note-Toolbar-Callouts) can now open menus, execute commands (without the need for a separate plugin), focus on folders, and show separators and breaks.
- Two methods are available for menus, commands, and focussing on folders:
  - _URIs_ (see examples below)
    - URIs don't appear to work in notes that were opened in a new window, as the focus is returned to the main window on click. If that's not suitable for you, use the next method.
  - _data elements_ (see examples below)
    - The `folder`, `menu`, and `command` types require empty links `()`, or Obsidian will open the link content instead.
    - Ensure the `<data>` element is closed (i.e., ends with a `/>`).
    - These links work in new windows opened from a note as well, unlike `obsidian://` URIs which seem to return focus to the main window.
  - Notes for both methods:
    - A message is displayed if the provided command, folder, or URI is invalid.
    - For commands, you can use either `command=` or `commandid=`  as both work the same.
    - Commands must be encoded (e.g., `:` becomes `%3A`) or you can surround the Obsidian URL with `<` and `>`.
    - For menus, you can pass in the toolbar name or the UUID, as shown below.
- For separators use `- <hr>` 
- For breaks use `- <br>`

*Example using URIs, data elements, and separators + breaks:*

```markdown
> [!note-toolbar]
> - [Command URI](obsidian://note-toolbar?command=workspace%3Atoggle-pin)
> - [Folder URI](obsidian://note-toolbar?folder=Demos)
> - <hr>
> - [Menu URI](obsidian://note-toolbar?menu=Tools)
> - [Menu URI with ID](obsidian://note-toolbar?menu=7fb30215-d92c-43ce-8158-b79096672bd1)
> - <br>
> - [Command]()<data data-ntb-command="workspace:toggle-pin"/>
> - [Folder]()<data data-ntb-folder="Demos"/>
> - <hr>
> - [Menu]()<data data-ntb-menu="Tools"/>
> - [Menu with ID]()<data data-ntb-menu="7fb30215-d92c-43ce-8158-b79096672bd1"/>
```

*Demo:*

<img src="https://github.com/user-attachments/assets/3aa59734-484c-40cd-8b48-26af750b39dd" width="700"/>

### Other features

- A new _space between items_ (desktop and mobile: `between`, mobile only: `mbtwn`) style has been added that spreads items out edge to edge.
- URI parameter to open settings:
  - `obsidian://note-toolbar?toolbarsettings`: opens the current toolbar's settings
  - `obsidian://note-toolbar?toolbarsettings=TOOLBARNAME` opens the settings of the provided toolbar (if it exists)

## Improvements 🎉

- Renaming your toolbars will no longer break mappings and menu items that use that toolbar.
  - Be aware that if you used the toolbar's name in a note's property to display that toolbar, it will _not_ be updated. Check the toolbar's Usage section for a link to search the configured property for the toolbar's name, to find the notes you'll need to update.
  - UUIDs are now added to toolbars and items as their DOM element IDs, which, as a side benefit, allows you to apply your own [click handlers](https://github.com/chrisgurney/obsidian-note-toolbar/discussions/110) if you desire.
  - This is also fixes a problem where toolbar deletions were also deleting toolbars of the same name (particularly those with empty names, which are allowed).
- Reorganized [Style Settings](https://github.com/chrisgurney/obsidian-note-toolbar/wiki/Style-Settings-plugin-support) options and added icons (📱) next to mobile-specific settings, to help make settings easier to find.
- Floating buttons are available on desktop:
  - are larger in size than on mobile; and 
  - have Style Settings available for background color, hover color, icon color, padding, position, icon size, icon stroke width, and border radius.
- The opacity of toolbars and floating buttons are _slightly_ reduced when the editor is not in focus, to provide some user feedback that an additional click is required (matching the behavior of Obsidian's tab and header title when they dim). See [#69](https://github.com/chrisgurney/obsidian-note-toolbar/issues/69).
- Various cosmetic improvements in settings.

## Fixes

- Settings: There was a UI issue that made it sometimes impossible to select a style from the dropdown on mobile.
- Callout list bullets no longer show in Reading mode.
- Middle clicking should now open file items in a new tab. (Fixed in v1.9.4)
- If you use [Iconize](https://github.com/FlorianWoelki/obsidian-iconize) to add icons to your notes (Callout links), they will now be correctly vertically aligned with the text. (Fixed in v1.9.4)

**Comment by chrisgurney** [comment]
@Steven-AA @FeralFlora @FelipeRearden @Moyf @MrComputron _and anybody else reading this_ I would love some help with testing this beta release (see the notes). 💙

**Comment by Moyf** [comment]
## Data json
I use git to manage my vault and after updating the plugin, it seems that my `data.json` file hasn't change:
![image](https://github.com/user-attachments/assets/1464a135-54d2-4cc6-9e6a-9c7cf4e32286)

---

Oh, after restarting Obsidian, it changed!
![image](https://github.com/user-attachments/assets/f5500dc8-02de-467c-9b91-1be9227c4f89)

1. The toolbar name  became uid
2. Link changed to uuid and new uuid property was added
![image](https://github.com/user-attachments/assets/37f0290c-7a5d-4153-84d7-f9f1833b69ea)

> 🤔 Maybe add a notice for users to **remember restart Obsidian after updating plugin** ?

---

Annnd.. The item group seems works well!

![image](https://github.com/user-attachments/assets/bbd6c6a5-7bf8-4148-8d84-00d4430713d0)

---

btw, the `See [beta-17]` in article lost its link 😂

**Reply by chrisgurney** [reply]
Thank you!

- Those diffs help a lot. Looks like it migrated properly!
- Restarting Obsidian seems to only be necessary when trying out betas with BRAT.
  - I'd be curious if you have this same experience with other plugins, or if there's something else I need to do.
  - I do have instructions in the [readme](https://github.com/chrisgurney/obsidian-note-toolbar?tab=readme-ov-file#installation) for betas, and show the version number in the Settings to provide feedback to the user.
  - I'm open to ideas if there's other things I can do. I've seen other plugins show popups on update (e.g., QuickAdd), but I've decided to not to be intrusive with messages, notifications, etc. if I can help it, unless absolutely necessary.
- Link to [beta-17](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.10.0-beta-17) fixed. Thanks!

**Reply by Moyf** [reply]
> Thank you!
> 
> * Those diffs help a lot. Looks like it migrated properly!
> * Restarting Obsidian seems to only be necessary when trying out betas with BRAT.
>   
>   * I'd be curious if you have this same experience with other plugins, or if there's something else I need to do.
>   * I do have instructions in the [readme](https://github.com/chrisgurney/obsidian-note-toolbar?tab=readme-ov-file#installation) for betas, and show the version number in the Settings to provide feedback to the user.
>   * I'm open to ideas if there's other things I can do. I've seen other plugins show popups on update (e.g., QuickAdd), but I've decided to not to be intrusive with messages, notifications, etc. if I can help it, unless absolutely necessary.
> * Link to [beta-17](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.10.0-beta-17) fixed. Thanks!

**About restartng**:
It might be related to the plugin itself? I can see it in *Tasks*'s update log:
![image](https://github.com/user-attachments/assets/c1e7648d-2aaa-4d1b-8507-17e9e3939f2a)
